vm/devices/net/net_mana/src/lib.rs

```diff
@@ -1118,17 +1122,155 @@ impl<T: DeviceBacking> ManaQueue<T> {
+                                            data_len,
+                                            "failed to bounce buffer coalesce extend"
+                                        );
+                                        // Drop the packet
+                                        return Ok(None);
+                                    }
+                                };
+                            (buf, gpa)
+                        };
+                        let mut next = buf.as_slice_from(gpa as usize); // CRASH HERE as GPA was 8742924288 :(
+                        for seg in &segment_vector {
+                            let len = seg.len as usize;
+                            self.guest_memory
+                                .read_to_atomic(seg.gpa as u64, &next[..len])?;
+                            next = &next[len..];
+                        }
+                        tracing::error!(
+                            "ERIK writing buffer to sgl_index {:?} data_len {:?}",
+                            sgl_index,
+                            data_len
+                        );
+                        sgl[sgl_index] = Sge {
+                            address: gpa,
+                            mem_key: self.mem_key,
+                            size: data_len as u32,
+                        };
+                        sgl_index += 1;
+                        segments_remaining -= segment_vector.len();
+                        segment_index += segment_vector.len();
+                        bounced_len_with_padding_total += data_len;
+                        bounce_buffer = Some(buf);
+                        bounce_buffer_next_gpa = Some(gpa + data_len as u64);
+                    }
+                }
+                if segments_remaining >= hardware_segment_limit {
+                    tracelimit::error_ratelimited!(
+                        segments_remaining,
+                        hardware_segment_limit,
+                        "Failed to bounce buffer the packet too many segments"
+                    );
+                    // Drop the packet, no need to free bounce buffer
+                    return Ok(None);
+                }
+                tracing::error!(
+                    "ERIK Finished Coalescing segment_index {:?} sgl_index {:?} segments_remaining {:?}",
+                    segment_index,
+                    sgl_index,
+                    segments_remaining
+                );
+                let sgl = &mut sgl[..segments_remaining];
+                for (tail, sge) in segments[segment_index..].iter().zip(&mut sgl[sgl_index..]) {
+                    *sge = Sge {
+                        address: self.guest_memory.iova(tail.gpa).unwrap(),
+                        mem_key: self.mem_key,
+                        size: tail.len,
+                    };
+                }
+                sgl
+            };
+
+            if let Some(buf) = bounce_buffer {
```

Comment:
the other option is to loop through the segments and compute how much bounce buffer is needed and allocate once from the bounce buffer. That way you only need to deal with the allocation failure once. Makes it easier to deal with the error, but does add the cost of an extra loop of the segments.